File: src/main/java/org/digma/intellij/plugin/jaegerui/model/incoming/Span.java
package org.digma.intellij.plugin.jaegerui.model.incoming;

import com.fasterxml.jackson.annotation.*;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import org.digma.intellij.plugin.common.CodeObjectsUtil;
import org.jetbrains.annotations.*;

@JsonIgnoreProperties(ignoreUnknown = true)
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY)
@JsonSerialize
public record Span(
        @JsonProperty(value = "id", required = true) String id,
        @JsonProperty(value = "name", required = true) String name,
        @JsonProperty(value = "instrumentationLibrary", required = true) String instrumentationLibrary,
        //TODO: spanCodeObjectId - add 'required = true' once backend version is out
        @JsonProperty(value = "spanCodeObjectId") String spanCodeObjectId,
        @JsonProperty(value = "methodCodeObjectId") String methodCodeObjectId, // can be null or empty
        // deprecated - use methodCodeObjectId
        @JsonProperty("function") String function,
        // deprecated - use methodCodeObjectId
        @JsonProperty("namespace") String namespace
) {

    @JsonCreator(mode = JsonCreator.Mode.PROPERTIES)
    public Span {
        //nothing to do , for jackson only
    }

    @NotNull
    public String spanId() {
        if (spanCodeObjectId != null) { // TODO: once spanCodeObjectId is required remove this if
            return CodeObjectsUtil.stripSpanPrefix(spanCodeObjectId);
        }
        // backward compatibility ,but note its not accurate since span name might be different than span name at spanCodeObjectId
        return CodeObjectsUtil.createSpanId(instrumentationLibrary, name);
    }

    @Nullable
    public String methodId() {
        if (methodCodeObjectId != null) {
            return CodeObjectsUtil.stripMethodPrefix(methodCodeObjectId);
        }
        if (function != null && namespace != null) {
            return CodeObjectsUtil.createMethodCodeObjectId(namespace, function);
        }
        return null;
    }

}

File: src/main/kotlin/org/digma/intellij/plugin/ui/jaegerui/model/Span.kt
package org.digma.intellij.plugin.ui.jaegerui.model

import com.fasterxml.jackson.annotation.*
import com.fasterxml.jackson.databind.annotation.JsonSerialize
import org.digma.intellij.plugin.common.CodeObjectsUtil
import java.beans.ConstructorProperties

@JsonIgnoreProperties(ignoreUnknown = true)
data class Span
@JsonCreator(mode = JsonCreator.Mode.PROPERTIES)
@ConstructorProperties(
    "id",
    "name",
    "instrumentationLibrary",
    "spanCodeObjectId",
    "methodCodeObjectId",
    "function",
    "namespace"
)
constructor(
    val id: String,
    val name: String,
    val instrumentationLibrary: String,
    val spanCodeObjectId: String?,
    // can be null or empty
    val methodCodeObjectId: String?,
    // deprecated - use methodCodeObjectId
    val function: String?,
    // deprecated - use methodCodeObjectId
    val namespace: String?
) {

    fun spanId(): String {
        return if (spanCodeObjectId != null) {
            // TODO: once spanCodeObjectId is required, remove this if
            CodeObjectsUtil.stripSpanPrefix(spanCodeObjectId)
        } else {
            // backward compatibility (not fully accurate)
            CodeObjectsUtil.createSpanId(instrumentationLibrary, name)
        }
    }

    fun methodId(): String? {
        return when {
            methodCodeObjectId != null -> CodeObjectsUtil.stripMethodPrefix(methodCodeObjectId)
            function != null && namespace != null -> CodeObjectsUtil.createMethodCodeObjectId(namespace, function)
            else -> null
        }
    }
}